### PR TITLE
Bugs/708 improve time value function

### PR DESCRIPTION
**Pull request by NAccuardi**

### CODE REVIEW

File: EPPlus/FormulaParsing/Excel/Functions/DateTime/TimeStringParser.cs [hunk 1]
```diff
@@ -26,10 +26,19 @@
 using System.Text.RegularExpressions;
 
 namespace OfficeOpenXml.FormulaParsing.Excel.Functions.DateTime
-{
+{   /// <summary>
+	/// TimeStringParser is used in TimeValue. It handles parsing the input and returning a double of just the time.
+	/// 
+	/// Note: The double that this will return is slightly diffrent from from the one you would get from TimeValue in Excel. the diffrence is that in excel if you do timevalue("26:00") you would get a value that is
```
Comment (emdelaney) [inline EPPlus/FormulaParsing/Excel/Functions/DateTime/TimeStringParser.cs]: This example is reproduced in the unit tests, so we can remove it from this header comment :)

File: EPPlus/FormulaParsing/Excel/Functions/DateTime/TimeValue.cs [hunk 2]
```diff
@@ -1,34 +1,62 @@
...
 using OfficeOpenXml.FormulaParsing.ExpressionGraph;
 
 namespace OfficeOpenXml.FormulaParsing.Excel.Functions.DateTime
 {
 	/// <summary>
-	/// Simple implementation of TimeValue function, just using .NET built-in
-	/// function System.DateTime.TryParse, based on current culture
+	/// Simple implementation of TimeValue function.
+	/// 
+	/// This makes use of the TimeStringParser class.
```
Comment (emdelaney) [inline EPPlus/FormulaParsing/Excel/Functions/DateTime/TimeValue.cs]: This is unnecessary since it can be easily deduced from the code, and can be removed :)

File: EPPlus/FormulaParsing/Excel/Functions/DateTime/TimeValue.cs [hunk 3]
```diff
@@ -1,34 +1,62 @@
-﻿using System.Collections.Generic;
+﻿/*******************************************************************************
```
Comment (emdelaney) [inline EPPlus/FormulaParsing/Excel/Functions/DateTime/TimeValue.cs]: Good catch, I always forget to add this :D I've added a task to the User Story to ensure we didn't miss it anywhere else (we don't need to fix it right now).

File: EPPlus/FormulaParsing/Excel/Functions/DateTime/TimeValue.cs [hunk 4]
```diff
@@ -1,34 +1,62 @@
...
 	public class TimeValue : ExcelFunction
 	{
 		public override CompileResult Execute(IEnumerable<FunctionArgument> arguments, ParsingContext context)
 		{
 			if (ValidateArguments(arguments, 1) == false)
 				return new CompileResult(eErrorType.Value);
+
 			var dateString = ArgToString(arguments, 0);
+
 			return Execute(dateString);
```
Comment (emdelaney) [inline EPPlus/FormulaParsing/Excel/Functions/DateTime/TimeValue.cs]: this.Execute()

File: EPPlus/FormulaParsing/Excel/Functions/DateTime/TimeValue.cs [hunk 5]
```diff
@@ -1,34 +1,62 @@
...
-	/// function System.DateTime.TryParse, based on current culture
+	/// Simple implementation of TimeValue function.
+	/// 
+	/// This makes use of the TimeStringParser class.
 	/// </summary>
 	public class TimeValue : ExcelFunction
 	{
 		public override CompileResult Execute(IEnumerable<FunctionArgument> arguments, ParsingContext context)
 		{
 			if (ValidateArguments(arguments, 1) == false)
```
Comment (emdelaney) [inline EPPlus/FormulaParsing/Excel/Functions/DateTime/TimeValue.cs]: this.ValidateArguments

File: EPPlus/FormulaParsing/Excel/Functions/DateTime/TimeValue.cs [hunk 6]
```diff
@@ -1,34 +1,62 @@
...
 				return new CompileResult(eErrorType.Value);
+
 			var dateString = ArgToString(arguments, 0);
```
Comment (emdelaney) [inline EPPlus/FormulaParsing/Excel/Functions/DateTime/TimeValue.cs]: this.argtostring

File: EPPlus/FormulaParsing/Excel/Functions/DateTime/TimeValue.cs [hunk 7]
```diff
@@ -1,34 +1,62 @@
...
-				 CreateResult(ExcelErrorValue.Create(eErrorType.Value), DataType.ExcelError);
-		}
+			var testString = dateString;
 
-		private double GetTimeValue(System.DateTime result)
-		{
-			return (int)result.TimeOfDay.TotalSeconds == 0 ? 0d : result.TimeOfDay.TotalSeconds / (3600 * 24);
+			var parser = new TimeStringParser();
+			var resultAfterParse = parser.Parse(testString);
+			var resultDecimalsOnly = resultAfterParse - System.Math.Truncate(resultAfterParse);
```
Comment (emdelaney) [inline EPPlus/FormulaParsing/Excel/Functions/DateTime/TimeValue.cs]: This can probably just be Math.Truncate, with a using system.Math; directive.

File: EPPlus/FormulaParsing/Excel/Functions/DateTime/TimeValue.cs [hunk 8]
```diff
@@ -1,34 +1,62 @@
...
-				 CreateResult(GetTimeValue(result), DataType.Date) :
-				 CreateResult(ExcelErrorValue.Create(eErrorType.Value), DataType.ExcelError);
-		}
+			var testString = dateString;
 
-		private double GetTimeValue(System.DateTime result)
-		{
-			return (int)result.TimeOfDay.TotalSeconds == 0 ? 0d : result.TimeOfDay.TotalSeconds / (3600 * 24);
+			var parser = new TimeStringParser();
+			var resultAfterParse = parser.Parse(testString);
```
Comment (emdelaney) [inline EPPlus/FormulaParsing/Excel/Functions/DateTime/TimeValue.cs]: We should rewrite this to be an internal static bool TryParse(string timeCandidate, out double time) method. I can help you out with this if necessary :) Then we'd check the result of that method, and return the error if we fail to parse, or proceed as normal if we can parse successfully.

File: EPPlusTest/FormulaParsing/Excel/Functions/DateTimeFunctions/TimeValueTests.cs [hunk 9]
```diff
@@ -30,6 +30,7 @@
 using OfficeOpenXml;
 using OfficeOpenXml.FormulaParsing.Excel.Functions.DateTime;
 
+
```
Comment (emdelaney) [inline EPPlusTest/FormulaParsing/Excel/Functions/DateTimeFunctions/TimeValueTests.cs]: This extra line can go away :)

File: EPPlusTest/FormulaParsing/Excel/Functions/DateTimeFunctions/TimeValueTests.cs [hunk 10]
```diff
@@ -40,11 +41,170 @@ public class TimeValueTests : DateTimeFunctionsTestBase
 		public void TimeValueWithInvalidArgumentReturnsPoundValue()
 		{
 			var func = new TimeValue();
-
```
Comment (emdelaney) [inline EPPlusTest/FormulaParsing/Excel/Functions/DateTimeFunctions/TimeValueTests.cs]: Hughes would be proud <3

File: EPPlusTest/FormulaParsing/Excel/Functions/DateTimeFunctions/TimeValueTests.cs [hunk 11]
```diff
@@ -40,11 +41,170 @@ public class TimeValueTests : DateTimeFunctionsTestBase
...
+			var input2 = "01:00";
+			var input3 = "26:00";
+			var input4 = "02:00";
+
+			var result1 = function.Execute(FunctionsHelper.CreateArgs(input1), this.ParsingContext);
+			var result2 = function.Execute(FunctionsHelper.CreateArgs(input2), this.ParsingContext);
+			var result3 = function.Execute(FunctionsHelper.CreateArgs(input3), this.ParsingContext);
+			var result4 = function.Execute(FunctionsHelper.CreateArgs(input4), this.ParsingContext);
+
+			Assert.AreEqual(0.041666667, System.Math.Round(result1.ResultNumeric,9));
```
Comment (emdelaney) [inline EPPlusTest/FormulaParsing/Excel/Functions/DateTimeFunctions/TimeValueTests.cs]: Add spaces before the nines that don't have them in this file so they all look the same :)

File: EPPlusTest/FormulaParsing/Excel/Functions/DateTimeFunctions/TimeValueTests.cs [hunk 12]
```diff
@@ -40,11 +41,170 @@ public class TimeValueTests : DateTimeFunctionsTestBase
...
+			var input1 = "13:00 pm";
+			var input2 = "1:00 pm";
+			var input3 = "16:00 pm";
+			var input4 = "4:00 pm";
+
+			var result1 = function.Execute(FunctionsHelper.CreateArgs(input1), this.ParsingContext);
+			var result2 = function.Execute(FunctionsHelper.CreateArgs(input2), this.ParsingContext);
+			var result3 = function.Execute(FunctionsHelper.CreateArgs(input3), this.ParsingContext);
+			var result4 = function.Execute(FunctionsHelper.CreateArgs(input4), this.ParsingContext);
+
```
Comment (emdelaney) [inline EPPlusTest/FormulaParsing/Excel/Functions/DateTimeFunctions/TimeValueTests.cs]: Fun observation: 13:00 pm returns #VALUE! in excel, so we should probably do the same here.

File: EPPlusTest/FormulaParsing/Excel/Functions/TimeStringParserTests.cs [hunk 13]
```diff
@@ -31,6 +61,8 @@ public void CanParseShouldHandleValid12HourPatterns()
 			Assert.IsTrue(parser.CanParse("9:12:55 PM"), "Could not parse 9:12:55 PM");
 			Assert.IsTrue(parser.CanParse("7 AM"), "Could not parse 7 AM");
 			Assert.IsTrue(parser.CanParse("4:12 PM"), "Could not parse 4:12 PM");
+			Assert.IsTrue(parser.CanParse("1/1/2017 7 AM"), "Could not parse 1/1/2017 7 AM");
+			Assert.AreEqual(0.291666667, Math.Round(parser.Parse("1/1/2017 7 AM"), 9));
 		}
```
Comment (emdelaney) [inline EPPlusTest/FormulaParsing/Excel/Functions/TimeStringParserTests.cs]: I would add a unit test for each weird edge case you did in the previous class. I'd also consider deleting this class entirely and making this parser method a ConvertUtil method, like ConvertUtil.TryParseTimeString().

File: EPPlus/FormulaParsing/Excel/Functions/DateTime/TimeStringParser.cs (hunk 1 above)
Comment (NAccuardi) [inline EPPlus/FormulaParsing/Excel/Functions/DateTime/TimeStringParser.cs]: ✔️

File: EPPlus/FormulaParsing/Excel/Functions/DateTime/TimeValue.cs (hunk 3 above)
Comment (NAccuardi) [inline EPPlus/FormulaParsing/Excel/Functions/DateTime/TimeValue.cs]: ✔️

File: EPPlus/FormulaParsing/Excel/Functions/DateTime/TimeValue.cs (hunk 2 above)
Comment (NAccuardi) [inline EPPlus/FormulaParsing/Excel/Functions/DateTime/TimeValue.cs]: ✔️

File: EPPlus/FormulaParsing/Excel/Functions/DateTime/TimeValue.cs (hunk 5 above)
Comment (NAccuardi) [inline EPPlus/FormulaParsing/Excel/Functions/DateTime/TimeValue.cs]: ✔️

File: EPPlusTest/FormulaParsing/Excel/Functions/TimeStringParserTests.cs (hunk 13 above)
Comment (emdelaney) [inline EPPlusTest/FormulaParsing/Excel/Functions/TimeStringParserTests.cs]: (And comprehensively test that new method, of course.)

File: EPPlus/FormulaParsing/Excel/Functions/DateTime/TimeValue.cs (hunk 6 above)
Comment (NAccuardi) [inline EPPlus/FormulaParsing/Excel/Functions/DateTime/TimeValue.cs]: ✔️

File: EPPlus/FormulaParsing/Excel/Functions/DateTime/TimeValue.cs (hunk 4 above)
Comment (NAccuardi) [inline EPPlus/FormulaParsing/Excel/Functions/DateTime/TimeValue.cs]: ✔️

File: EPPlusTest/FormulaParsing/Excel/Functions/DateTimeFunctions/TimeValueTests.cs (hunk 9 above)
Comment (NAccuardi) [inline EPPlusTest/FormulaParsing/Excel/Functions/DateTimeFunctions/TimeValueTests.cs]: ✔️

File: EPPlusTest/FormulaParsing/Excel/Functions/DateTimeFunctions/TimeValueTests.cs (hunk 11 above)
Comment (NAccuardi) [inline EPPlusTest/FormulaParsing/Excel/Functions/DateTimeFunctions/TimeValueTests.cs]: Must of missed these ones on my once over. ✔️

File: EPPlusTest/FormulaParsing/Excel/Functions/TimeStringParserTests.cs (hunk 13 above)
Comment (NAccuardi) [inline EPPlusTest/FormulaParsing/Excel/Functions/TimeStringParserTests.cs]: I like this Idea. I am going to work on that after I fix the other items.

File: EPPlus/FormulaParsing/Excel/Functions/DateTime/TimeValue.cs (hunk 7 above)
Comment (NAccuardi) [inline EPPlus/FormulaParsing/Excel/Functions/DateTime/TimeValue.cs]: We did end up having to use system.math.truncate insted of math.truncate because OfficeOpenXml also has a math function that was causing issues. ✔️

File: EPPlus/FormulaParsing/Excel/Functions/DateTime/TimeValue.cs (hunk 8 above)
Comment (NAccuardi) [inline EPPlus/FormulaParsing/Excel/Functions/DateTime/TimeValue.cs]: For now I created the TryParseDateStringToDouble method directly in the TimeValue class. If we find we need it more places we can move it to the convertUtil. ✔️

File: EPPlusTest/FormulaParsing/Excel/Functions/DateTimeFunctions/TimeValueTests.cs (hunk 12 above)
Comment (NAccuardi) [inline EPPlusTest/FormulaParsing/Excel/Functions/DateTimeFunctions/TimeValueTests.cs]: This is giving me a ton of problems and uncovering more issues.

File: EPPlusTest/FormulaParsing/Excel/Functions/TimeStringParserTests.cs (hunk 13 above)
Comment (NAccuardi) [inline EPPlusTest/FormulaParsing/Excel/Functions/TimeStringParserTests.cs]: I added some test cases. ✔️

File: EPPlusTest/FormulaParsing/Excel/Functions/DateTimeFunctions/TimeValueTests.cs (hunk 12 above)
Comment (NAccuardi) [inline EPPlusTest/FormulaParsing/Excel/Functions/DateTimeFunctions/TimeValueTests.cs]: This bug is gonna stay for a bit.

File: EPPlus/FormulaParsing/Excel/Functions/DateTime/TimeStringParser.cs [hunk 14]
```diff
@@ -94,11 +94,13 @@ private double Parse12HourTimeString(string input)
 		{
 			string dayPart = string.Empty;
 			dayPart = input.Substring(input.Length - 2, 2);
```
Comment (emdelaney) [inline EPPlus/FormulaParsing/Excel/Functions/DateTime/TimeStringParser.cs]: You can chain the ToLower() call after the substring here if you'd like.

File: EPPlus/FormulaParsing/Excel/Functions/DateTime/TimeStringParser.cs [hunk 15]
```diff
@@ -57,6 +56,7 @@ private void ValidateValues(int hour, int minute, int second)
 			{
 				throw new FormatException("Illegal value for minute: " + minute);
 			}
+			
 		}
 
 		public virtual double Parse(string input)
```
Comment (emdelaney) [inline EPPlus/FormulaParsing/Excel/Functions/DateTime/TimeStringParser.cs]: Can get get rid of the Regex.IsMatch(input, RegEx12) call entirely?

Comment (emdelaney) [inline EPPlus/FormulaParsing/Excel/Functions/DateTime/TimeStringParser.cs]: Also, this should use TimeStringParser.RegEx12/24, etc.

File: EPPlus/FormulaParsing/Excel/Functions/DateTime/TimeStringParser.cs [hunk 16]
```diff
@@ -94,11 +94,13 @@ private double Parse12HourTimeString(string input)
...
 			dayPart = input.Substring(input.Length - 2, 2);
+			dayPart = dayPart.ToLower();
 			int hour;
 			int minute;
 			int second;
 			GetValuesFromString(input, out hour, out minute, out second);
-			if (dayPart == "PM") hour += 12;
+			if (dayPart == "pm" & hour !=12) hour += 12;
+			if (dayPart == "am" & hour == 12) hour = 0;
 			ValidateValues(hour, minute, second);
```
Comment (emdelaney) [inline EPPlus/FormulaParsing/Excel/Functions/DateTime/TimeStringParser.cs]: this.Validate, this.GetSerialNumber, etc

File: EPPlus/FormulaParsing/Excel/Functions/DateTime/TimeValue.cs [hunk 17]
```diff
@@ -24,39 +24,52 @@
...
+			var dateString = this.ArgToString(arguments, 0);
 
-			return Execute(dateString);
+			return this.Execute(dateString);
 		}
 
 		internal CompileResult Execute(string dateString)
 		{
-			var testString = dateString;
+			//var testString = dateString;
```
Comment (emdelaney) [inline EPPlus/FormulaParsing/Excel/Functions/DateTime/TimeValue.cs]: These commented-out lines can go away.

File: EPPlus/FormulaParsing/Excel/Functions/DateTime/TimeValue.cs [hunk 18]
```diff
@@ -24,39 +24,52 @@
...
+
+			//var parser = new TimeStringParser();
+			double resultAfterParse = 0.0;
+
+			TryParseDateStringToDouble(dateString, out resultAfterParse);
```
Comment (emdelaney) [inline EPPlus/FormulaParsing/Excel/Functions/DateTime/TimeValue.cs]: this.TryParse... 

You can also declare the double in-line, as in `out double resultAfterParse`.

File: EPPlus/FormulaParsing/Excel/Functions/DateTime/TimeValue.cs [hunk 19]
```diff
@@ -24,39 +24,52 @@
...
 
-			var parser = new TimeStringParser();
-			var resultAfterParse = parser.Parse(testString);
 			var resultDecimalsOnly = resultAfterParse - System.Math.Truncate(resultAfterParse);
 	
 			return resultAfterParse != -1 ?		//The '-1' is used to throw an error if an invalid input is supplied.
 				 CreateResult(resultDecimalsOnly, DataType.Decimal) :
 				 CreateResult(ExcelErrorValue.Create(eErrorType.Value), DataType.ExcelError);
 		}
+		private bool TryParseDateStringToDouble(string dateString, out double result )
```
Comment (emdelaney) [inline EPPlus/FormulaParsing/Excel/Functions/DateTime/TimeValue.cs]: There's an extra space after result that can be removed.

File: EPPlusTest/FormulaParsing/Excel/Functions/DateTimeFunctions/TimeValueTests.cs [hunk 20]
```diff
@@ -134,58 +137,66 @@ public void TimeValueIsNormal12HourClock()
...
-			var input3 = "12:00 pm";
-			var input4 = "12:00 am";
-			var input6 = "1:00 pm";
-			var input7 = "13:00";
+			var input1 = "00:00:00 AM";
+			var input2 = "00:01:32 AM";
+			var input3 = "12:00 PM";
+			var input4 = "12:00 AM";
+			var input6 = "1:00 PM";
+			//var input7 = "13:00 PM"; // I am putting this on ice for now. Excel Returns a #Value, EPP Retuns a value.
```
Comment (emdelaney) [inline EPPlusTest/FormulaParsing/Excel/Functions/DateTimeFunctions/TimeValueTests.cs]: I'd leave this test case in there but not commented out (leave the descriptive comment though), then if its behavior changes we know. I'd assert the current values you get back.

File: EPPlusTest/FormulaParsing/Excel/Functions/DateTimeFunctions/TimeValueTests.cs (hunk 12 above)
Comment (emdelaney) [inline EPPlusTest/FormulaParsing/Excel/Functions/DateTimeFunctions/TimeValueTests.cs]: Sounds good.

File: EPPlusTest/FormulaParsing/Excel/Functions/TimeStringParserTests.cs [hunk 21]
```diff
@@ -62,6 +65,9 @@ public void CanParseShouldHandleValid12HourPatterns()
 			Assert.IsTrue(parser.CanParse("7 AM"), "Could not parse 7 AM");
 			Assert.IsTrue(parser.CanParse("4:12 PM"), "Could not parse 4:12 PM");
 			Assert.IsTrue(parser.CanParse("1/1/2017 7 AM"), "Could not parse 1/1/2017 7 AM");
+			Assert.IsTrue(parser.CanParse("1.17.2011 2:00 am"), "Could not parse 1.17.2011 2:00 am");
```
Comment (emdelaney) [inline EPPlusTest/FormulaParsing/Excel/Functions/TimeStringParserTests.cs]: I'd throw a note/comment in here that this is different from Excel, which does not gracefully handle this case.

File: EPPlus/FormulaParsing/Excel/Functions/DateTime/TimeValue.cs (hunk 18 above)
Comment (NAccuardi) [inline EPPlus/FormulaParsing/Excel/Functions/DateTime/TimeValue.cs]: That is a pretty cool trick. ✔️

File: EPPlusTest/FormulaParsing/Excel/Functions/DateTimeFunctions/TimeValueTests.cs (hunk 20 above)
Comment (NAccuardi) [inline EPPlusTest/FormulaParsing/Excel/Functions/DateTimeFunctions/TimeValueTests.cs]: I added a bit more description to the comment ans made it assert the values it returns for now. ✔️

File: EPPlusTest/FormulaParsing/Excel/Functions/DateTimeFunctions/TimeValueTests.cs (hunk 10 above)
Comment (NAccuardi) [inline EPPlusTest/FormulaParsing/Excel/Functions/DateTimeFunctions/TimeValueTests.cs]: ✔️

File: EPPlus/FormulaParsing/Excel/Functions/DateTime/TimeStringParser.cs (hunk 14 above)
Comment (NAccuardi) [inline EPPlus/FormulaParsing/Excel/Functions/DateTime/TimeStringParser.cs]: ✔️

File: EPPlusTest/FormulaParsing/Excel/Functions/DateTimeFunctions/TimeValueTests.cs (hunk 12 above)
Comment (NAccuardi) [inline EPPlusTest/FormulaParsing/Excel/Functions/DateTimeFunctions/TimeValueTests.cs]: ✔️

File: EPPlus/FormulaParsing/Excel/Functions/DateTime/TimeStringParser.cs (hunk 16 above)
Comment (NAccuardi) [inline EPPlus/FormulaParsing/Excel/Functions/DateTime/TimeStringParser.cs]: I think I got all of them.  ✔️

File: EPPlus/FormulaParsing/Excel/Functions/DateTime/TimeStringParser.cs (hunk 15 above)
Comment (NAccuardi) [inline EPPlus/FormulaParsing/Excel/Functions/DateTime/TimeStringParser.cs]: ✔️

File: EPPlusTest/FormulaParsing/Excel/Functions/DateTimeFunctions/TimeValueTests.cs [hunk 22]
```diff
@@ -167,10 +167,11 @@ public void TimeValueIsNormal12HourClock()
 
 		}
 	
-		/* This test Case is being put off for now. 
+		  
 		[TestMethod]
-		public void TimeValueTimesInTheForm1300PmShouldNotWork()
+		public void TimeValueTimesInTheForm1300PmShouldNotWork()//This test Case is being put off for now.
 		{
+			//This test Case is being put off for now. The correct Functionality would have result 1 and 3 returning a #Value.
```
Comment (emdelaney) [inline EPPlusTest/FormulaParsing/Excel/Functions/DateTimeFunctions/TimeValueTests.cs]: I'd maybe phrase this comment as "Note: In Excel, the argument 13:00 PM would return #VALUE!."

File: EPPlusTest/FormulaParsing/Excel/Functions/DateTimeFunctions/TimeValueTests.cs [hunk 23]
```diff
@@ -167,10 +167,11 @@ public void TimeValueIsNormal12HourClock()
 
 		}
 	
-		/* This test Case is being put off for now. 
+		  
 		[TestMethod]
-		public void TimeValueTimesInTheForm1300PmShouldNotWork()
+		public void TimeValueTimesInTheForm1300PmShouldNotWork()//This test Case is being put off for now.
```
Comment (emdelaney) [inline EPPlusTest/FormulaParsing/Excel/Functions/DateTimeFunctions/TimeValueTests.cs]: I'd eliminate this part of the comment, and also rename this method to TimeValueTimeOfTheForm1300Pm so we don't have to change it if/when we fix the behavior.
